### PR TITLE
Add nonblocking calls to usart

### PR DESCRIPTION
**Pull request by kungjohan**
Add nonblocking calls to check the status of the usart.
Instead of busy waiting you can do something else waiting to transmit next byte or waiting to receive data.

### CODE REVIEW

File: lib/stm32/common/usart_common_f124.c [hunk 1]
```diff
@@ -65,6 +65,24 @@ uint16_t usart_recv(uint32_t usart)
...
+ * the next data word.
+ *
+ * @param[in] usart unsigned 32 bit. USART block register address base @ref
+ * usart_reg_base
+ * @returns boolean: transmit data buffer is ready to accept the next data word
+ */
+
+bool usart_is_send_ready(uint32_t usart)
+{
+	/* Wait until the data has been transferred into the shift register. */
```
Comment (karlp) [inline lib/stm32/common/usart_common_f124.c]: copy/paste error?

File: lib/stm32/f0/usart.c [hunk 2]
```diff
@@ -191,6 +191,23 @@ uint8_t usart_recv(uint32_t usart)
...
+ * the next data word.
+ *
+ * @param[in] usart unsigned 32 bit. USART block register address base @ref
+ * usart_reg_base
+ * @returns boolean: transmit data buffer is ready to accept the next data word
+ */
+
+bool usart_is_send_ready(uint32_t usart)
+{
+	/* Wait until the data has been transferred into the shift register. */
```
Comment (karlp) [inline lib/stm32/f0/usart.c]: Same here

**Comment by karlp** [comment]
Seems generally reasonable.  I have no objections to this.

**Comment by karlp** [comment]
Oh, Please at least update the f3 file too, to at least keep the stm32 line consistent.  "soon" the f0, f3 and l0 peripherals will all get a common core pulled out, but until then you'll need to modify them all.  (except L0 doesn't exist just yet, so you're lucky there ;)

**Comment by kungjohan** [comment]
Thanks for the review, updated this.

**Comment by danielinux** [comment]
@kungjohan Thank you very much for this. We merged this change to our fork back in November and we strongly believe that such calls should be present in the unified UART driver for all platforms. Ref: https://git.io/voQGk

**Comment by karlp** [comment]
TODO (I will do in merge if no-one gets there first)

- squish to single commit
- modify all the wait_ blocking functions to use the new is_ asynch functions, rather than duplicating the checks

**Comment by OpenUAS** [comment]
Merge appreciated...

**Comment by karlp** [comment]
@OpenUAS you can see the TODO if you "want to get there first"  making less for me to do is always a good way of making it easier for things to happen.  Rebasing it to be current too.

**Pull request closed without merging**